feat: 히스토리 api ts rest 적용

nyj001012 opened the pull request:
### 개요
히스토리 API에 ts-rest 적용하여 리팩토링

### 작업 사항
- router
- controller
- service
- error

### 비고
1. #600 
2. book Id 안 쓰는거 같은데 없애면 안되나
3. api 나누기 (유저 권한에 따라 등) - who로 구별하기 번거로움

### 스크린샷 (optional)

nyj001012 committed:
feat: histories 라우터에 추가

histories 라우터에 추가

nyj001012 committed:
feat: histories controller 추가

histories controller 추가

nyj001012 committed:
feat: impl 추가

impl 추가

nyj001012 committed:
feat: histories 전용 type 파일 추가

histories 전용 type 파일 추가

nyj001012 committed:
feat: histories service 추가

histories service 추가

nyj001012 committed:
feat: 권한 없음 에러 추가

권한 없음 에러 추가

nyj001012 committed:
feat: 권한 없음 에러 오브젝트 추가

권한 없음 에러 오브젝트 추가

nyj001012 committed:
feat: meta 스키마 추가

meta 스키마 추가

scarf005 commented:
https://github.com/orgs/jiphyeonjeon-42/discussions/7#discussioncomment-6581705

scarf005 commented on backend/src/v2/shared/types.ts:
```diff
@@ -1,4 +1,14 @@
+  itemsPerPage: z.number().nonnegative(),
+  totalPages: z.number().nonnegative(),
+  currentPage: z.number().nonnegative(),
+});
```
contracts/ 패키지에서 사용하면 프론트와 백 모두에서 사용할수 있을것 같아용

nyj001012 replied:
metaSchema로 추가했습니다 :)

scarf005 commented on backend/src/v2/shared/types.ts:
```diff
@@ -1,4 +1,14 @@
+
+export type Meta = z.infer<typeof meta>;
+export const meta = z.object({
+  totalItems: z.number().nonnegative(),
```
https://github.com/jiphyeonjeon-42/backend/blob/cc1ec7b7e569302abb46e1a6eed7ac94c03f0e48/contracts/src/shared.ts#L3-L4

contracts의 positiveInt를 사용해보면 어떨까요?

nyj001012 replied:
좋습니다!

nyj001012 committed:
style: eslint comma 추가

eslint comma 추가

nyj001012 committed:
feat: contract에 histories 추가

contract에 histories 추가

nyj001012 committed:
feat: contract에 histories 명세 추가

contract에 histories 명세 추가

nyj001012 committed:
feat: contract에 histories에서 사용하는 query string 스키마 추가

contract에 histories에서 사용하는 GET 메서드들의 query string 스키마 추가

nyj001012 committed:
feat: contract에 공용으로 사용할 권한 없음 에러 스키마 추가

contract에 공용으로 사용할 권한 없음 에러 스키마 추가

nyj001012 committed:
feat(histories): histories controller 추가

- 전체 대출 기록 조회하는 controller 추가
- 나의 대출 기록 조회하는 controller 추가

nyj001012 committed:
feat(histories): histories service 추가

- 전체 대출 기록 조회하는 service 추가
- 나의 대출 기록 조회하는 service 추가

nyj001012 committed:
feat(histories): histories 조회 시의 query string schema 변경

histories 조회 시의 query string schema 구조 변경

nyj001012 committed:
feat(histories): 라우터 분기 자세하게 나눔

- histories의 라우터 분기를 '나의 기록 조회', '전체 기록 조회'로 나눔

nyj001012 commented:
@scarf005
```
{
  "issues": [
    {
      "code": "invalid_type",
      "expected": "number",
      "received": "string",
      "path": [
        "page"
      ],
      "message": "Expected number, received string"
    },
    {
      "code": "invalid_type",
      "expected": "number",
      "received": "string",
      "path": [
        "limit"
      ],
      "message": "Expected number, received string"
    }
  ],
  "name": "ZodError"
}
```
page와 limit가 string으로 들어가요! help...

nyj001012 committed:
feat(histories): histories service index.ts 추가

histories service index.ts 추가

scarf005 committed:
fix: 쿼리 파라미터를 json으로 파싱

참고:
https://ts-rest.com/docs/core/#query-parameters
https://ts-rest.com/docs/express/#options

nyj001012 committed:
feat(histories): 결과 반환 시, literal이 아닌 아이템 리스트 반환

결과 반환 시, z.literal 조회 성공 문구 대신 데이터베이스 조회한 결과 반환하게 함

nyj001012 committed:
style(histories): import 형식 변경

VHistories 가져올 때 중괄호 이용

nyj001012 committed:
feat(histories): histories 조회 결과 200 시, 스키마 추가

histories를 조회하여 200 스테이터스 일 때 반환하는 스키마 추가

nyj001012 committed:
feat(histories): 검색 조건 callsign 추가

callsign으로도 검색할 수 있도록 조건 추가 및 query undefined 체크 추가

nyj001012 committed:
feat(histories): meta 스키마 추가

meta 스키마 공용 스키마로 추가

nyj001012 committed:
feat(histories): id 타입 변경

id 타입을 string -> number로 변경

nyj001012 committed:
feat(histories): meta 스키마 positiveInt로 변경

z.number()...를 positiveInt로 변경

nyj001012 committed:
feat(histories): 서비스 함수 반환 타입 변경

서비스 함수 반환 타입 오브젝트 형식으로 변경

nyj001012 committed:
feat(histories): getMyHistories 반환값 변경

getMyHistories 반환값 데이터베이스 조회 결과로 변경

nyj001012 committed:
feat(histories): 서비스 함수 반환 타입 변경

서비스 함수들 반환 타입 변경

nyj001012 committed:
chore(histories): 서비스 파일 분리 및 컨트롤러 파일 분리

서비스 파일과 서비스 구현 파일, 컨트롤러 파일과 컨트롤러 구현 파일 분리

scarf005 commented on contracts/src/histories/schema.ts:
```diff
@@ -0,0 +1,32 @@
+      image: z.string(),
+      createdAt: z.string(),
+      returnedAt: z.string(),
+      updatedAt: z.date(),
```
`VHistories`의 `createdAt` 필드의 타입은 `Date`이나, `createdAt`의 타입이 `string`이어서 타입 오류가 나고 있습니다. 하지만 typeorm 쿼리에서 날짜를 문자열 형식으로 반환하기 때문에, 이와 같은 경우에는 `coerce`를 사용하는 것을 추천드립니다.

https://github.com/colinhacks/zod/discussions/879

scarf005 commented on backend/src/v2/histories/controller/controller.ts:
```diff
@@ -0,0 +1,50 @@
+    .otherwise(() => ({
+      status: 200,
+      body: result,
+    } as const));
```
![image](https://github.com/jiphyeonjeon-42/backend/assets/54838975/57705ad0-6b0f-4918-8953-0809c2b458fb)

반환값이 `UnauthorizedError | { items: [...], meta: { ... } }` 타입이기 때문에, `UnauthorizedError`를 예상하지 못한 ts-rest에서 오류를 내고 있습니다. 다음을 참고하시면 타입을 좁힐 때 도움이 될 것 같습니다.

https://github.com/gvergnaud/ts-pattern#withpattern-handler
https://github.com/gvergnaud/ts-pattern#exhaustive-otherwise-and-run

scarf005 committed:
Merge branch 'develop' into 598-히스토리-api-ts-rest-적용

scarf005 committed:
fix: 스키마에서 date와 string 모두 허용

scarf005 committed:
fix: 패턴 매칭에서 타입이 좁혀지지 않던 문제

scarf005 committed:
fix: users 경로 숨김

scarf005 committed:
fix: stocks 타입 오류 수정

scarf005 reviewed (approved):
![image](https://github.com/jiphyeonjeon-42/backend/assets/54838975/78898028-0235-48e0-9e2b-b472cfc70883)
![image](https://github.com/jiphyeonjeon-42/backend/assets/54838975/c3a2cecd-a491-40b8-a3f1-bba69987a697)

LGTM! 시간을 끌어 죄송합니다...
마지막에 #663 을 위한 타입 수정도 조금 추가했습니다
